File: vetix/plugins/base64_exec.py
import re

from vetix.plugin import Issue, Plugin, Severity
from vetix.utils.utils import get_relative_path


class Base64ExecPlugin(Plugin):
    """Detect base64-decoded payloads piped directly into a shell (sh/bash)."""

    name = "Base64 Exec"

    _EXEC_RE = re.compile(
        r"base64\s+(?:-d|--decode)[^|]*?\|\s*(ba)?sh",
        re.IGNORECASE,
    )

    def scan(self, skill_dir: str, file_path: str, content: str) -> list[Issue]:
        issues: list[Issue] = []
        relative_path = get_relative_path(file_path, skill_dir)

        for match in self._EXEC_RE.finditer(content):
            line = content.count("\n", 0, match.start()) + 1
            issues.append(Issue(
                name="Base64 command piped to shell",
                severity=Severity.CRITICAL,
                category="Remote Execution",
                description=(
                    f"Base64-decoded command is piped into a shell ({match.group(0)})."
                ),
                file_path=relative_path,
                line=line,
                audit_required=True,
            ))
        return issues

File: internal/plugin/base64_exec.go
package plugin

import (
	"fmt"
	"regexp"
)

var base64ExecRe = regexp.MustCompile(`(?i)base64\s+(?:-d|--decode)[^|]*?\|\s*(ba)?sh`)

type Base64ExecPlugin struct{}

func (Base64ExecPlugin) Meta() Meta {
	return Meta{
		ID:          "base64_exec",
		Name:        "Base64 Command Execution",
		Description: "Detects a Base64-decoded command piped straight into a shell, a common disguise for remote code execution.",
	}
}

func (Base64ExecPlugin) Scan(skillDir, filePath, content string) []Issue {
	var issues []Issue
	rel := relativePath(filePath, skillDir)
	for _, m := range base64ExecRe.FindAllStringIndex(content, -1) {
		issues = append(issues, Issue{
			Name:          "Base64 command piped to shell",
			Severity:      SeverityCritical,
			Category:      CatRemoteExecution,
			Description:   fmt.Sprintf("Base64-decoded command is piped into a shell (%s).", content[m[0]:m[1]]),
			FilePath:      rel,
			Line:          lineOf(content, m[0]),
			AuditRequired: true,
		})
	}
	return issues
}
